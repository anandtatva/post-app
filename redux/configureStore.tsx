import { combineReducers, createStore } from "redux";
import { persistStore, persistReducer } from "redux-persist";
import AsyncStorage from "@react-native-async-storage/async-storage";
import PostListSlice from "./reducer/post-list-slice";

const persistConfig = {
    key: "post-list",
    storage: AsyncStorage
}

const rootReducer = combineReducers({
    posts: PostListSlice
})

const persistReducers = persistReducer(persistConfig, rootReducer);

const store = createStore(persistReducers);

const persistor = persistStore(store);

export {
    store,
    persistor
}