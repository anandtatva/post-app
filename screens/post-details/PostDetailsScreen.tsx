import { Text, View, StyleSheet } from "react-native"
import { commonStyle, colors } from "../../utils";
const index = "Details"
const PostDetailsScreen = ({ route }) => {
    const { item = {} } = route?.params || {}
    return (
        <View style={styles.container}>
            <View style={styles.detailsContainer}>
                <Text style={styles.title} testID={`Title`}>Title</Text>
                <Text style={styles.text} numberOfLines={1} testID={`Title-${index}`}>{item.title}</Text>
            </View>
            <View style={styles.detailsContainer}>
                <Text style={styles.title} testID={`Author`}>Author</Text>
                <Text style={styles.text} testID={`Author-${index}`}>{item.author}</Text>
            </View>
            <View style={styles.detailsContainer}>
                <Text style={styles.title} testID={`URL`}>URL</Text>
                <Text style={styles.text} testID={`URL-${index}`}>{item.url} {item.url}{item.url}</Text>
            </View>
            <View style={styles.detailsContainer}>
                <Text style={styles.title} testID={`Created-At`}>Created At</Text>
                <Text style={styles.text} testID={`Created-At-${index}`}>{item.created_at}</Text>
            </View>
        </View>
    )
}

export default PostDetailsScreen;

const styles = StyleSheet.create({
    detailsContainer: {
        paddingVertical: 5,
    },
    container: {
        flex: 1,
        backgroundColor: colors.primary,
        paddingHorizontal: 10,
        paddingVertical: 10
    },
    title: {
        textAlign: "center",
        paddingBottom: 5,
        fontWeight: "bold",
        color: colors.secondary,
    },
    text: {
        textAlign: "center",
        paddingLeft: 5,
        color: colors.secondary,
        opacity: 0.8,
        fontWeight: "300",
    }
})