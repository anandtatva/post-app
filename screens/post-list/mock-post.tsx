import { PostType } from "../../types/post-type";

export const mockPostList: PostType[] = [
    { title: "New Song1", url: "http://new-song1.com", created_at: "1", author: "Jack1" },
    { title: "New Song2", url: "http://new-song2.com", created_at: "2", author: "Jack2" }
]