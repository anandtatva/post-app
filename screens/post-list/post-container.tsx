import { Text, TouchableOpacity, View, StyleSheet } from "react-native"
import { PostType } from "../../types/post-type";
import { FC } from "react";
import { colors } from "../../utils";


interface PostContainerProps {
    index: number;
    item: PostType;
    onNext: (item: PostType) => void;
}
const PostContainerView: FC<PostContainerProps> = ({
    item, index, onNext
}) => {
    const handleNext = () => { onNext(item) }
    return (
        <TouchableOpacity
            onPress={handleNext}
            testID={`Post-Row-${index}`}
            style={styles.conatiner}
        >
            <View style={styles.detailsContainer}>
                <Text style={styles.title} testID={`Title`}>Title</Text>
                <Text style={styles.text} numberOfLines={1} testID={`Title-${index}`}>{item.title}</Text>
            </View>
            <View style={styles.detailsContainer}>
                <Text style={styles.title} testID={`Author`}>Author</Text>
                <Text style={styles.text} numberOfLines={1} testID={`Author-${index}`}>{item.author}</Text>
            </View>
            <View style={styles.detailsContainer}>
                <Text style={styles.title} testID={`URL`}>URL</Text>
                <Text style={styles.text} numberOfLines={1} testID={`URL-${index}`}>{item.url} {item.url}{item.url}</Text>
            </View>
            <View style={styles.detailsContainer}>
                <Text style={styles.title} testID={`Created-At`}>Created At</Text>
                <Text style={styles.text} numberOfLines={1} testID={`Created-At-${index}`}>{item.created_at}</Text>
            </View>
        </TouchableOpacity>
    )
}

export default PostContainerView;

const styles = StyleSheet.create({
    detailsContainer: {
        paddingVertical: 2,
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "space-between",
    },
    conatiner: {
        flex: 1,
        backgroundColor: colors.primary,
        paddingHorizontal: 10,
        paddingVertical: 10
    },
    title: {
        flex: 2,
        fontWeight: "bold",
        color: colors.secondary

    },
    text: {
        flex: 4,
        paddingLeft: 5,
        color: colors.secondary,
        fontWeight: "300",
    }
})