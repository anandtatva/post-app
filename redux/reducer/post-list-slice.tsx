import { createSlice } from "@reduxjs/toolkit";
import { PostType } from "../../types/post-type";
import { mockPostList } from "../../screens/post-list/mock-post";
interface PostListStateProps {
    totalPage: number;
    currentPage: number;
    postList: PostType[];
}

const initialState: PostListStateProps = {
    postList: [],
    currentPage: 0,
    totalPage: 0

}
const PostListSlice = createSlice({
    name: "post-list-slice",
    initialState,
    reducers: {
        setCurrentPage: (state, action) => {
            state.currentPage = action.payload
        },
        setTotalPage: (state, action) => {
            state.totalPage = action.payload
        },
        setPostListPage: (state, action) => {
            state.postList = action.payload
        },
        setAllData: (state, action) => {
            state.postList = action.payload.currentPage === 0 ? action.payload.postList : [...state.postList, ...action.payload.postList]
            state.totalPage = action.payload.totalPage
            state.currentPage = action.payload.currentPage
        }
    }
})

export const { setAllData, setCurrentPage, setTotalPage, setPostListPage } = PostListSlice.actions;

export default PostListSlice.reducer;