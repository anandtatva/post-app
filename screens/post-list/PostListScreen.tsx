import { Text, View, FlatList, StyleSheet } from "react-native"
import { mockPostList } from "./mock-post";
import PostContainerView from "./post-container";
import { PostType } from "../../types/post-type";
import { useSelector, useDispatch } from "react-redux";
import { colors, strings, commonStyle, showAlert } from "../../utils";
import { getPostAPIs } from "../../services/api";
import { setTotalPage, setAllData } from "../../redux/reducer/post-list-slice";
import { useEffect } from "react";

const PostListScreen = ({ navigation }) => {
    const postList = useSelector(state => state.posts)
    const dispatch = useDispatch();
    useEffect(() => {
        _loadPostList(0);
    }, [])
    const _loadPostList = async (page: number) => {
        const resp = await getPostAPIs(page);
        if (resp.status) {
            const { data, currentPage, totalPage } = resp;
            dispatch(setAllData({
                totalPage,
                postList: data,
                currentPage
            }))
            if (totalPage > currentPage + 1) {
                setTimeout(() => {
                    _loadPostList(page + 1)
                }, 10000);
            }
        } else showAlert(resp.message)
    }
    const renderItem = ({ item, index }: { item: PostType; index: number; }) => (
        <PostContainerView
            onNext={(item: PostType) => { navigation.navigate("PostDetailsScreen", { item }) }}
            item={item}
            index={index}
        />
    )
    return (
        <FlatList
            ListHeaderComponent={() => <View style={{
                paddingVertical: 10,
            }}>
                <Text style={styles.emptyText} testID="Current-Total-Post">Current Total Post:- {postList.postList.length}</Text>
                <Text style={styles.emptyText} testID="Current-Page-Total-Page">Current Page:- {postList?.currentPage + 1} Total Page:-  {postList?.totalPage}</Text>
            </View>}
            ItemSeparatorComponent={() => <View style={styles.separator} />}
            data={postList?.postList || []}
            keyExtractor={(_, index) => String(index)}
            renderItem={renderItem}
            ListEmptyComponent={() => (
                <Text style={styles.emptyText} testID="Empty-Post-List" >{strings.emptyPostList}</Text>
            )}
        />
    )
}

export default PostListScreen;

const styles = StyleSheet.create({
    separator: {
        borderBottomWidth: 1.5,
        borderColor: colors.gray
    },
    emptyText: {
        fontWeight: "bold",
        textAlign: "center"
    }
})