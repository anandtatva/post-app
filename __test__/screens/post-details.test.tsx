import { cleanup, fireEvent, render, waitFor } from "@testing-library/react-native";
import { Provider } from "react-redux";
import { store } from "../../redux/configureStore";
import { setAllData } from "../../redux/reducer/post-list-slice";
import PostDetailsScreen from "../../screens/post-details/PostDetailsScreen";
import { mockPostList } from "../../screens/post-list/mock-post";


describe("Screen: Post Details", () => {
    beforeEach(() => {
        store.dispatch(setAllData({
            currentPage: 0,
            totalPage: 0,
            postList: []
        }))
    })
    afterEach(() => {
        cleanup()
        jest.clearAllMocks()
    })
    it("renders correctly", () => {
        const route = {
            params: {
                item: mockPostList[0]
            }
        }
        const tree = render(
            <Provider store={store}>
                <PostDetailsScreen route={route} />
            </Provider>
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    it("should be show correct details of any post", () => {
        const item = mockPostList[0];
        const route = {
            params: {
                item
            }
        }
        const { getByTestId } = render(
            <Provider store={store}>
                <PostDetailsScreen route={route} />
            </Provider>
        )
        const title = getByTestId("Title-Details");
        const url = getByTestId("URL-Details");
        const author = getByTestId("Author-Details");
        const createdAt = getByTestId("Created-At-Details");

        expect(title.children[0]).toBe(item.title)
        expect(url.children[0]).toBe(item.url)
        expect(createdAt.children[0]).toBe(item.created_at)
        expect(author.children[0]).toBe(item.author)
    })
})