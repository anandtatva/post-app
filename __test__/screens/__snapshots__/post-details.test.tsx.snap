// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen: Post Details renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000",
      "flex": 1,
      "paddingHorizontal": 10,
      "paddingVertical": 10,
    }
  }
>
  <View
    style={
      Object {
        "paddingVertical": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontWeight": "bold",
          "paddingBottom": 5,
          "textAlign": "center",
        }
      }
      testID="Title"
    >
      Title
    </Text>
    <Text
      numberOfLines={1}
      style={
        Object {
          "color": "#FFF",
          "fontWeight": "300",
          "opacity": 0.8,
          "paddingLeft": 5,
          "textAlign": "center",
        }
      }
      testID="Title-Details"
    >
      New Song1
    </Text>
  </View>
  <View
    style={
      Object {
        "paddingVertical": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontWeight": "bold",
          "paddingBottom": 5,
          "textAlign": "center",
        }
      }
      testID="Author"
    >
      Author
    </Text>
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontWeight": "300",
          "opacity": 0.8,
          "paddingLeft": 5,
          "textAlign": "center",
        }
      }
      testID="Author-Details"
    >
      Jack1
    </Text>
  </View>
  <View
    style={
      Object {
        "paddingVertical": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontWeight": "bold",
          "paddingBottom": 5,
          "textAlign": "center",
        }
      }
      testID="URL"
    >
      URL
    </Text>
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontWeight": "300",
          "opacity": 0.8,
          "paddingLeft": 5,
          "textAlign": "center",
        }
      }
      testID="URL-Details"
    >
      http://new-song1.com
       
      http://new-song1.com
      http://new-song1.com
    </Text>
  </View>
  <View
    style={
      Object {
        "paddingVertical": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontWeight": "bold",
          "paddingBottom": 5,
          "textAlign": "center",
        }
      }
      testID="Created-At"
    >
      Created At
    </Text>
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontWeight": "300",
          "opacity": 0.8,
          "paddingLeft": 5,
          "textAlign": "center",
        }
      }
      testID="Created-At-Details"
    >
      1
    </Text>
  </View>
</View>
`;
